### ISSUE TITLE
[bug] Windows conan_toolchain.cmake broken using CMake "CodeBlocks - NMake Makefiles" generator with CLion

### ISSUE THREAD

**Issue by drmacdon** [conan-io/conan#10041]
### Environment Details
  * Operating System+version: Windows 10 x64
  * Compiler+version: Visual Studio 2017 and 2019
  * Conan version: 1.42.1
  * Python version: 3.9.2

### Steps to reproduce (Include if Applicable)

CLion invokes the CMake generator with -G "CodeBlocks - NMake Makefiles" and the current code to _get_generator_platform() is detecting a Windows Visual Studio build and returns "x64" here https://github.com/conan-io/conan/blob/develop/conan/tools/cmake/toolchain.py#L567.

This results in the inclusion of the property CMAKE_GENERATOR_PLATFORM here https://github.com/conan-io/conan/blob/develop/conan/tools/cmake/toolchain.py#L503 giving the additional conan_toolchain.cmake line:

```
set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE) 
```

This fails because it is apparently not a supported option for the CMake NMake Makefiles. 

If the set CMAKE_GENERATOR_PLATFORM line is not present in the generated conan_toolchain.cmake file, CMake is happy and is able to configure the project.

### Logs (Executed commands with output) (Include/Attach if Applicable)

The error CLion generates is: 
```
CMake Error at CMakeLists.txt:2 (project):
  Generator

    NMake Makefiles

  does not support platform specification, but platform

    x64

  was specified.
```

**Comment by franramirez688** [comment]
Hi @drmacdon 

I think the problem may be that you're not telling conan the generator via `CMakeToolchain` or `[conf]`.
Check if you're defining one of these two options to declare the generator:

1- In your current conan `profile`, you can declare something like:
```
[settings]
...

[conf]
tools.cmake.cmaketoolchain:generator="CodeBlocks - NMake Makefiles"
```

2- In your `conanfile.py`, invoke `CMakeToolchain()` and pass the generator:
```python
from conans import ConanFile
from conan.tools.cmake import CMakeToolchain

class YourApp(ConanFile):
    # ...

    def generate(self):
        tc = CMakeToolchain(self, generator="CodeBlocks - NMake Makefiles")
        # ...
        tc.generate()
```

If you're already using one of those, then we'll have to dig deeper into the issue.

**Comment by drmacdon** [comment]
Hi @franramirez688 

Thank you for the quick response. I have tested both your suggestions and confirm they both work. 

I had been trying to use CONAN_CMAKE_GENERATOR_PLATFORM which isn't leveraged in this code path but didn't realize the option to specify as a property to CMakeToolchain or even know the [conf] section in the profile existed. Both are valid and reasonable solutions with associated tradeoffs. Namely, maintaining multiple profiles if building in Clion or not. 

If the check was deferred until usage time, I wouldn't need to do anything and be less likely to cause others problems.

ie. Having [toolchain.py](https://github.com/conan-io/conan/blob/develop/conan/tools/cmake/toolchain.py#L503) guard against the output of that property if a CMAKE_GENERATOR exists, which occurs when CMake is launched with the -G option as it is in the case of CLion or someone manually overriding from the CLI. In either case the combination of the two properties is invalid.

```python
        {% if generator_platform %}      
        # NMake Makefiles does not support CMAKE_GENERATOR_PLATFORM specification 
        if(NOT CMAKE_GENERATOR STREQUAL "NMake Makefiles")
          set(CMAKE_GENERATOR_PLATFORM "{{ generator_platform }}" CACHE STRING "" FORCE)
        endif()  
        {% endif %}
```

Thanks!

**Comment by memsharded** [comment]
> I had been trying to use CONAN_CMAKE_GENERATOR_PLATFORM w

Yes, in general, all configuration env-vars will be deprecated and superseded by ``conf`` alternatives.

>  Namely, maintaining multiple profiles if building in Clion or not.

You don't need to maintain multiple profiles, you can specify the ``-c config_name=config-value`` in the command line directly

> If the check was deferred until usage time, I wouldn't need to do anything and be less likely to cause others problems.

The thing is that the generator is used for other things. For example it can affect the project ``layout()``, which is important to implement local developer flows and allow ``editable`` packages. If it was necessary strictly for the cmake invocation, it could be possible, but it isn't. Then it is needed at ``generate()`` time, that is, defined in the ``toolchain``.

**Comment by drmacdon** [comment]
> Yes, in general, all configuration env-vars will be deprecated and superseded by conf alternatives.

This makes sense, just didn't realize it was deprecated when I was trying that option. 

> You don't need to maintain multiple profiles, you can specify the -c config_name=config-value in the command line directly

I didn't know that option existed either, and while that sort of works, it is technically broken somewhere (IDE Conan plugin?) due to escaping logic. 

With ```--build missing -c tools.cmake.cmaketoolchain:generator="NMake Makefiles"```
```bash
conan install C:/Users/drmacdon/Development/test -if=C:\Users\drmacdon\Development\test\cmake-build-debug -pr=default-debug-2019 --build missing -c tools.cmake.cmaketoolchain:generator=\"NMake Makefiles\"
...
conan install: error: unrecognized arguments: Makefiles"
```

While ```--build missing -c tools.cmake.cmaketoolchain:generator=NMake\ Makefiles```
```bash
conan install C:/Users/drmacdon/Development/test -if=C:\Users\drmacdon\Development\test\cmake-build-debug -pr=default-debug-2019 --build missing -c tools.cmake.cmaketoolchain:generator=NMake\ Makefiles
...
conan install: error: unrecognized arguments: Makefiles
```

Since I could not find a way to propagate the proper property through, I gave up and used "anything" which makes it sufficient to avoid the toolchain.py generation of the field. Conan presently assumes that I want the CMAKE_GENERATOR_PLATFORM value if there is nothing in the generator field... which is equally invalid. At least my suggested guard had the benefit of being always technically correct and consistent within CMake and broke no one any more than they might already have been.

Presently, my recipe is building with CLion, CLI and VSCode on OSX & Windows. I did have to provide my own layout because none of the provided ones seemed to work for me in a general recipe sense. Presently, this is what I'm using and why:

```python
    def layout(self):
        cmake_layout(self)   # Something to start with in case I miss something

        # The default cmake_layout uses "." to identify the source location but when using VS Code
        # the "." is where VS Code was launch from which is problematic. The best way to do this
        # is to anchor relative to this conanfile. The build_type is left as-is instead of
        # dropping to lowercase because VS Code treats it as an environment variable when building
        # paths which would be impossible to match case. This allows the VS Code Conan Plugin to
        # define it's JSON profiles like so,
        #     "installArg": "--build missing --install-folder ${workspaceFolder}/cmake-build-${buildType}",
        source_path = os.path.dirname(__file__)
        build_path = os.path.join(source_path, "cmake-build-{}".format(self.settings.build_type))

        self.folders.source = source_path
        self.folders.build = build_path   # Required to build from arbitrary command line location
        self.folders.generators = os.path.join(self.folders.build, "conan")  # Expect to be relative to build location
        self.folders.imports = self.folders.build   # Want imports to go to present build as they used to
```

**Comment by franramirez688** [comment]
Hi @drmacdon 

You don't really need to put the `\` before the quotes. Try this:

```bash
conan install C:/Users/drmacdon/Development/test -if=C:\Users\drmacdon\Development\test\cmake-build-debug -pr=default-debug-2019 --build missing -c tools.cmake.cmaketoolchain:generator="NMake Makefiles"
```
I hope it works now 😁

**Comment by drmacdon** [comment]
> it is technically broken somewhere (IDE Conan plugin?) due to escaping logic.
>
> With --build missing -c tools.cmake.cmaketoolchain:generator="NMake Makefiles"
> 
> conan install C:/Users/drmacdon/Development/test -if=C:\Users\drmacdon\Development\test\cmake-build-debug -pr=default-debug-2019 --build missing -c tools.cmake.cmaketoolchain:generator=\\"NMake Makefiles\\"
> ...
> conan install: error: unrecognized arguments: Makefiles"

I think you misunderstand. While I can get it to work from the command line where I can control the quoting, as mentioned the IDE/Conan plugin is escaping and I don't have control over that.

My point was being technically correct using "NMake Makefiles" is broken on the IDE/plugin, while giving "anything" works because the *actual property value is irrelevant* to Conan because it is also broken in a different way as it *needs something/anything* to literally trick it into not setting a property that should not be set. Which was the point of this ticket. 

@memsharded highlighted that I could end up with problems not providing the eventual CMake generator information to Conan - although I presently don't appear to have a problem as the actual property value is irrelevant. That may be because I provided my own layout or am not using editable packages at present.

In the process I have learned lots of ways to configure that I was unaware of (thanks!), and the last of which would have been ideal except for the escaping problem. It is however, still a work around to trick Conan to do the right thing because of an incorrect/_unchecked assumption_ being made when generating the conan_toolchain.cmake. Those assumptions work flawlessly on a Mac and they work fine on Windows with VS Code (uses CMake generator of type "Visual Studio *") but fail on Windows with CLion (uses CMake generator "NMake Makefiles").

**Comment by memsharded** [comment]
So @drmacdon just to make sure I fully understand the issue: at the moment the main blocker is the CLion Conan plugin, that is not doing the right thing? Is it possible to work with CLion defining the NMake Makefiles, if not using the plugin? In that case, we would need to work on the plugin repo, maybe transfer this issue there?

**Comment by drmacdon** [comment]
At the moment, I have no blocker,  I'm using ```-c tools.cmake.cmaketoolchain:generator="anything"``` (literally).  That is an issue with double quotes and GeneralCommandLine and I'm not entirely certain if it is the plugin, the IDE, the terminal expansion, ```conan``` or a possible interaction of multiple of them.

With respect to this ticket, at this point, my only concern was to highlight that the ```toolchain.py``` was currently leaving **unchecked assumptions** in the generated code. Given the install and build phases happen distinctly and the ```conan_toolchain.cmake``` is being fed forward without an explicit generator, one was **implicitly assumed** which led to  my original problem and likely others in the future.
 
@memsharded suggested explicitly providing the toolchain was most appropriate. I suggested a code patch which would mitigate this present assumption where it is causing me grief as I assumed it was just an overlooked property combination. Another solution would be to assert the assumption providing information on how to fix if it is truly required. 

```cmake
if (NOT CMAKE_GENERATOR STREQUAL "{{ generator }}")
     message(FATAL_ERROR "Conan CMake generator '{{ generator }}' does not match CMake generator '${CMAKE_GENERATOR}'.\nSet tools.cmake.cmaketoolchain:generator='${CMAKE_GENERATOR}' during conan install.")
endif()     
```

**Comment by memsharded** [comment]
We don't want to introduce a mitigation that only partially addresses this, because the generator is actually needed for a bunch of things:

- To generate a ``vcvars`` file when needed:
```python
 # Generators like Ninja or NMake requires an active vcvars
        elif self.generator is not None and "Visual" not in self.generator:
            VCVars(self._conanfile).generate()
```

- generation of the ``ClangCL`` toolset for MSVC:
```python
        elif compiler == "clang":
            if generator and "Visual" in generator:
                if "Visual Studio 16" in generator:
                    return "ClangCL"
```

- Definition of the ``cl`` compiler:
```python
  if (generator is not None and "Ninja" in generator
                and (compiler is not None and "Visual" in compiler or compiler == "msvc")):
            compiler = "cl"
        else:
```

- Definition of the Intel compiler toolset:
```python
        elif compiler == "intel" and compiler_base == "Visual Studio" and "Visual" in generator:
            # TODO: This intel toolset needs to be validated too
            compiler_version = settings.get_safe("compiler.version")
            if compiler_version:
                compiler_version = compiler_version if "." in compiler_version else \
                    "%s.0" % compiler_version
                return "Intel C++ Compiler " + compiler_version
```

It is not a thing that can be delayed, or just checked at build time. It is really needed at ``generate()`` time to do the right thing for the toolchains. We had the approach you are suggesting before, and it didn't work, so we had to change it to request the CMake generator at ``generate()`` time. We could force the users to input always one generator. The thing is that it is not necessary for many cases. The default implicitly deduced generator (like deducing the "Visual Studio" generator for visual compiler is right in 90% of the cases. Users deviating from the defaults will need to provide the details, but it seems unnecessary to force all users to provide a generator.

**Comment by drmacdon** [comment]
That all makes sense and is reasonable. It is nice that Conan is able to guess the generator, providing a reasonable default and be correct 90% of the time - it's a little less clear on what it guessed, what is actually compatible and what is not. 

When Conan installs, it guesses and configures the default profile, we can view and edit the default profile if it was guessed wrongly. The generator is a bit more obscure, it is implicit and lives in code. 

Given that cmake allows the user to change, I would think it is worth while to check and at least warn that something is likely to cause problems for those of us unfortunate enough to be living in the 10%.

**Comment by memsharded** [comment]
Yes, that makes sense, lets try to add some checks (initially I think it would be better to warn than to error, to not accidentally block users because some bug in the checks)

**Comment by spectras** [comment]
It seems that on CMake-based projects, if Ninja is installed then **Visual Studio uses Ninja by default**.
This results in conan guess being wrong. At least on our systems (conan 1.42.2, visual studio 16.11.6, cmake 3.22).

Here is the call that Visual Studio does to generate cmake cache:
```
"C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\PROFESSIONAL\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Release" -DCMAKE_INSTALL_PREFIX:PATH="D:\project\out\install\x64-Release" -DCMAKE_TOOLCHAIN_FILE:FILEPATH="D:\project\out\build\x64-Release\conan_toolchain.cmake" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe" -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\PROFESSIONAL\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "D:\project" 2>&1"
```
This fails with: `Generator Ninja does not support platform specification, but platform x64 was specified.`

→ This is what happens when opening a CMake-based project in Visual Studio for the first time and trying to build it.

This is very frustrating due to lack of documentation. In fact, if it were not for this issue ticket, we would still be investigating.
With CMake being a first-class citizen in Visual Studio now, it would be great if the default configuration worked out of the box.

**Comment by memsharded** [comment]
Hi @spectras 

The problem is that it is still more common to generate the VS solution with ``cmake`` (Visual Studio generator) and then opening the generated .sln solution, than opening a CMake project directly from the IDE with the CMake integration.

It is impossible to Conan to define the toolchain with the same inputs for 2 completely different scenarios, with different CMake generators. So the above still holds, it is necessary for Conan to make something correct to know which CMake generator is going to be used. At the moment the ``Visual Studio`` default guess is still the least bad, as it keeps being the most  popular approach.

**Comment by spectras** [comment]
Hello @memsharded !
Yes. Although with Microsoft working hard on their native CMake support, this is likely to be more and more used. As this goes on, conan will fail by default on an out-of-the-box install of Visual Studio — with a hard-to-diagnose problem.

Making VisualStudio aware of conan would be perfect, but with Microsoft pushing vcpkg it is unlikely to happen. So making this scenario work from conan side sounds reasonable.

I understand the issues with making `conan_toolchain.cmake` generator-agnostic. As a general approach, it is not really doable. However, I believe this change could be implemented:
- If the detected generator is Visual Studio,
- then insert a generate-time check for Ninja that skips setting `CMAKE_GENERATOR_PLATFORM`.

That's all that is needed to make it work. Visual Studio already knows to add things that Ninja needs when calling CMake. In particular, no need for an active CVars (in fact, they should even be *avoided* in that case as they could interfere with VisualStudio settings).

Would that make sense to you? If so, I could look into making a PR of that at some point.

**Comment by sztomi** [comment]
I ran into this with the ninja generator. I don't think it's impossible to do this and @spectras 's suggestion is the exact right one. The generated toolchain file may  contain logic based on the current generator used and set only the compatible variables in all cases. This would be an especially desirable fix since in 2.0, all other CMake generators are deprecated if I'm not mistaken, so building with MSVC via anything other than msbuild is broken.

**Comment by memsharded** [comment]
Hi all,

It seems this ticket would need an update.

The problem that I see is that changing the generator do not only affect the toolchain. It affects many other parts like:
- The project layout(), which is necessary for editable and workspaces
- The CMake generated presets

So I still think that a single ``conan install`` that will handle different CMake generators is not possible, and that the CMake generator needs to be passed as ``conf`` in ``conan install`` time.

**Comment by memsharded** [comment]
Closing as staled, please open new tickets for any further question, thanks!
